provider/labels.go

```diff
@@ -18,119 +17,65 @@ import (
-			*paths = append(*paths, currentPath)
+	toUpdate := map[string][]resource.PropertyPath{}
 
-			// This field represents an aggregate of resource-level `labels` and provider-level `defaultLabels` fields.
-			// To avoid leaking secrets set via Input on those fields, we mark this field as unconditionally Secret.
-			info.Secret = tfbridge.True()
-		}
-		// Also set `effectiveLabels` to Secret.
-		if key == "effective_labels" {
-			info.Secret = tfbridge.True()
+	visitor := func(ctx tfbridge.PropertyVisitInfo) (tfbridge.PropertyVisitResult, error) {
+		if ctx.Root.PulumiToken() == "" { // Skip unmapped resources and functions
+			return tfbridge.PropertyVisitResult{}, nil
 		}
-
-		if m == nil {
-			return
+		path := ctx.SchemaPath()
+		key, ok := path[len(path)-1].(walk.GetAttrStep)
+		if !ok {
+			return tfbridge.PropertyVisitResult{}, nil
 		}
+		switch key.Name {
+		case "terraform_labels":
+			ctx.SchemaInfo().Name = "pulumiLabels"
 
-		switch elem := m.Elem().(type) {
-		case shim.Resource: // this is a nested object
-			obj := elem.Schema()
-			if info.Elem == nil {
-				info.Elem = new(tfbridge.SchemaInfo)
+			// We only apply this transform for resources
+			if root, ok := ctx.Root.(tfbridge.VisitResourceRoot); ok {
+				toUpdate[root.TfToken] = append(toUpdate[root.TfToken],
+					tfbridge.SchemaPathToPropertyPath(path, ctx.EnclosingSchemaMap(),
+						ctx.EnclosingSchemaInfoMap()))
 			}
 
-			if info.Elem.Fields == nil {
-				info.Elem.Fields = make(map[string]*tfbridge.SchemaInfo, obj.Len())
-			}
-			obj.Range(func(key string, m shim.Schema) bool {
-				i, ok := info.Elem.Fields[key]
-				if !ok {
-					i = new(tfbridge.SchemaInfo)
-					info.Elem.Fields[key] = i
-				}
-				pulumiKey := tfbridge.TerraformToPulumiNameV2(key, obj, info.Elem.Fields)
-				currentPath = append(currentPath, pulumiKey)
-				apply(key, m, i, paths, currentPath)
-				return true
-			})
-		case shim.Schema: // this is a map or array or a Set
-			if info.Elem == nil {
-				info.Elem = new(tfbridge.SchemaInfo)
-			}
-			currentPath = append(currentPath, "*")
-			apply("", elem, info.Elem, paths, currentPath)
-		}
-	}
-	// update takes a terraform schema map and a pulumi schema info map of a given resource, and calls `apply` to do
-	// necessary resource transformations.
-	// - Rename terraformLabels to pulumiLabels at schema generation time
-	// - Create a collection of resource property paths that have this pulumiLabels key amd return it.
-	update := func(s shim.SchemaMap, fields *map[string]*tfbridge.SchemaInfo) []resource.PropertyPath {
-		var paths []resource.PropertyPath
+			ctx.SchemaInfo().Secret = tfbridge.True()
 
-		if *fields == nil {
-			*fields = make(map[string]*tfbridge.SchemaInfo, s.Len())
+		// This field represents an aggregate of resource-level `labels` and provider-level `defaultLabels` fields.
+		// To avoid leaking secrets set via Input on those fields, we mark this field as unconditionally Secret.
+		case "effective_labels":
```

Comment:
Does this suffer from the performance issues stemming from the secretness like  tags in AWS?

Reply:
Only during tfgen time. During normal runtime it only touches the resources it needs to.